pkg/scheduler/framework/plugins/nodeaffinity/node_affinity.go

```diff
@@ -83,10 +88,51 @@ func (pl *NodeAffinity) EventsToRegister() []framework.ClusterEvent {
 }
 
 // PreFilter builds and writes cycle state used by Filter.
-func (pl *NodeAffinity) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) *framework.Status {
+func (pl *NodeAffinity) PreFilter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod) (*framework.PreFilterResult, *framework.Status) {
 	state := &preFilterState{requiredNodeSelectorAndAffinity: nodeaffinity.GetRequiredNodeAffinity(pod)}
 	cycleState.Write(preFilterStateKey, state)
-	return nil
+	affinity := pod.Spec.Affinity
+	if affinity == nil ||
+		affinity.NodeAffinity == nil ||
+		affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil ||
+		len(affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) == 0 {
+		return nil, nil
+	}
+
+	// Check if there is affinity to a specific node and return it.
+	terms := affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
+	var nodeNames sets.String
+	for _, t := range terms {
+		var termNodeNames sets.String
+		for _, r := range t.MatchFields {
```

Comment:
Is the plan to cover term.MatchExpressions in a follow-up?

Reply:
we can't cover it, unlike metadata.Name which is guaranteed to be unique, there is no guarantee that labels are unique and actually match the node name.

Reply:
```yaml
NodeSelectorTerms: []v1.NodeSelectorTerm{
    {
        MatchFields: []v1.NodeSelectorRequirement{
            {
                Key:      metav1.ObjectNameField,
                Operator: v1.NodeSelectorOpIn,
                Values:   []string{"node_1"},
            },
            {
                Key:      metav1.ObjectNameField,
                Operator: v1.NodeSelectorOpNotIn,
                Values:   []string{"node_1"},
            },
        },
    },
},
```

I contract a test case.  It returns `&{NodeNames: sets.String{"node_1": {}}},`. ` v1.NodeSelectorOpNotIn` is also needed to handle here.

Reply:
I would leave it to the filter stage, there is no practical performance gain from this, and we should try to keep the logic here as simple as simple as we can.